crashtracker/src/collector/crash_handler.rs

```diff
@@ -41,37 +248,52 @@ enum ReceiverType {
 static ALTSTACK_INIT: AtomicBool = AtomicBool::new(false);
 static OLD_HANDLERS: AtomicPtr<OldHandlers> = AtomicPtr::new(ptr::null_mut());
-static RECEIVER: AtomicPtr<ReceiverType> = AtomicPtr::new(ptr::null_mut());
 static METADATA: AtomicPtr<(CrashtrackerMetadata, String)> = AtomicPtr::new(ptr::null_mut());
 static CONFIG: AtomicPtr<(CrashtrackerConfiguration, String)> = AtomicPtr::new(ptr::null_mut());
-
-fn make_receiver(config: &CrashtrackerReceiverConfig) -> anyhow::Result<std::process::Child> {
-    // TODO: currently create the file in write mode.  Would append make more sense?
-    let stderr = if let Some(filename) = &config.stderr_filename {
-        File::create(filename)?.into()
-    } else {
-        Stdio::null()
-    };
-
-    let stdout = if let Some(filename) = &config.stdout_filename {
-        File::create(filename)?.into()
-    } else {
-        Stdio::null()
-    };
-
-    let receiver = Command::new(&config.path_to_receiver_binary)
-        .args(&config.args)
-        .envs(config.env.clone())
-        .stdin(Stdio::piped())
-        .stderr(stderr)
-        .stdout(stdout)
-        .spawn()
-        .context(format!(
-            "Unable to start process: {}",
-            config.path_to_receiver_binary
-        ))?;
-
-    Ok(receiver)
+static RECEIVER_CONFIG: AtomicPtr<CrashtrackerReceiverConfig> = AtomicPtr::new(ptr::null_mut());
+static RECEIVER_ARGS: AtomicPtr<PreparedExecve> = AtomicPtr::new(ptr::null_mut());
+
+fn make_receiver(config: &CrashtrackerReceiverConfig) -> anyhow::Result<Receiver> {
+    let stderr = open_file_or_quiet(config.stderr_filename.as_deref())?;
+    let stdout = open_file_or_quiet(config.stdout_filename.as_deref())?;
+
+    // Create anonymous Unix domain socket pair for communication
+    let (uds_parent, uds_child) = socket::socketpair(
+        socket::AddressFamily::Unix,
+        socket::SockType::Stream,
+        None,
+        socket::SockFlag::empty(),
+    )
+    .context("Failed to create Unix domain socket pair")
+    .map(|(a, b)| (a.into_raw_fd(), b.into_raw_fd()))?;
+
+    // We need to spawn a process without calling atfork handlers, since this is happening inside
+    // of a signal handler.  Moreover, preference is given to multiplatform-uniform solutions.
+    // Although `vfork()` is deprecated, the alternatives have limitations
+    // * `fork()` calls atfork handlers
+    // * There is no guarantee that `posix_spawn()` will not call `fork()` internally
+    // * `clone()`/`clone3()` are Linux-specific
+    // Accordingly, use `vfork()` for now
+    // NB -- on macos the underlying implementation here is actually `fork()`!  See the top of this
+    // file for details.
+    match unsafe { vfork() } {
+        0 => {
+            // Child (noreturn)
+            run_receiver_child(uds_parent, uds_child, stderr, stdout)
+        }
+        pid if pid > 0 => {
+            // Parent
+            run_receiver_parent(uds_parent, uds_child, stderr, stdout);
```

Comment:
Minor nit: I find that code confusing ... calling a function with 4 args, just uses one and then it's an one-liner.
```suggestion
            _ = close(uds_child);
```
Trivial.

Reply:
Agree, refactored